### PR TITLE
Refresh cookbook notebooks for TabPFN 3.5 and add rerun tooling

### PR DESCRIPTION
**Pull request by psinger-prior**
## Summary

TabPFN 3.5 became the default in `tabpfn` 9.0.0 and `tabpfn-client` 0.6.0 on 2026-09-15. All 11 notebooks whose saved outputs depend on the model were re-executed against 3.5, and prose was updated where the new numbers changed the conclusion. The rerun is now a script, so the next model release is one command instead of a manual pass through Colab. Along the way the reruns exposed three notebooks that no longer ran from a clean install, which are fixed here.

## What changed

```diff
 scripts/
+├── rerun_notebooks.py          # builds one uv env per notebook from its pip line, runs, converts
+├── _execute_notebook.py        # runs inside that env: nbclient, in-place write, failure copy
+└── colab_shim/google/colab/    # userdata.get(name) -> environment variable
+datasets/
+└── forecast_spare_parts_demand/spare_parts_daily.csv   # was generated locally, never committed
 notebooks/, markdowns/, visuals/   # 11 notebooks re-executed on 3.5, MDX and plots regenerated
```

- **Outputs refreshed** for quickstart, insurance, predictive distribution, XGBoost comparison, thinking mode, synthetic data, cache, spare-parts forecast, interpret results, decoder readout. Bayesian optimization was refreshed too but is reverted here for a separate review, since its results changed substantially.
- **Prose updated** where 3.5 changed the story: insurance (TabPFN Gini ~0.34 vs GLM ~0.31, single-stage log target now edges the two-stage model), XGBoost table label says 3.5. Thinking mode was reworked by its author on this branch with a grouped `group_col` example.
- **Notebooks that were broken on a fresh install**: synthetic data dropped `table-evaluator` and `dython` (table-evaluator pins scikit-learn 1.5.1, which forces `tabpfn-extensions` back to 0.1.3) in favor of pandas, seaborn, and scikit-learn plots; the forecast notebook reads its CSV from `datasets/` or the raw GitHub URL on Colab; thinking mode, interpret results, and decoder readout now install `matplotlib`, which they imported but only got for free on Colab.
- **Output noise removed**: estimator repr leaking the client's `sentry-trace` header, Styler reprs instead of tables, a shap progress line, and a pip-install log containing local paths in the time-series notebook.
- **Cache notebook** got an untimed warm-up fit so model loading is not counted in the no-cache timing.

## How it works

```text
rerun_notebooks.py  (repo env, stdlib only)
  for each slug
    select notebooks whose pip line installs tabpfn / tabpfn-client / tabpfn-extensions
    parse `pip install` lines -> --with specs
    mark install-only cells to skip
    uv run --no-project --refresh --exclude-newer <now> --python 3.12 --with ... \
        python _execute_notebook.py <nb> --skip-cell N --failed-copy .rerun/<slug>.failed.ipynb
      PYTHONPATH=scripts/colab_shim  TABPFN_NO_BROWSER=1  cwd=notebooks/
    convert_to_markdown.py --slug <slug>
```

The executor keeps notebook metadata, cell ids, and JSON indentation as found, merges spinner output the way Jupyter does, and writes the notebook only on success. Any failure leaves the file untouched and writes the partial run to `.rerun/`.

## Review guide

- Start with `scripts/rerun_notebooks.py`, then `scripts/_execute_notebook.py`. Everything else is either generated or a targeted notebook edit.
- The `--exclude-newer <now>` flag exists because a user-level uv `exclude-newer` setting silently hid the day-old 3.5 releases during development. It is harmless where that setting is absent.
- The runner deliberately does not add Colab's preinstalled packages. A notebook that imports something its pip line does not list fails here, and it also fails for anyone outside Colab, so the notebook is what gets fixed.
- Notebook diffs are large because outputs were regenerated. Prose changes worth reading: insurance intro, the four insurance markdown cells around the TabPFN results and takeaways, the synthetic-data comparison section.
- The rerun flow is deliberately not documented in the README. Releases are handled manually; the script's docstring covers usage.
- Two notebooks had one noisy stderr output removed by hand after adding a warning filter, without a full re-execution: insurance and interpret results. Numbers are untouched.
- Local-inference notebooks ran on a datacenter GPU, so their timings differ from the Colab numbers they replaced.

## Testing

- `uv run python scripts/validate.py --all` passes, `git diff --check` is clean.
- Runner tested end to end on the old release first, then used for the actual 3.5 refresh of all 11 notebooks, including the failure path (three notebooks failed on first run and left `.rerun/` copies).
- Every notebook and generated MDX scanned for home directories, cache paths, kernel temp paths, IPs, tokens, and sentry headers. None remain.
- Not tested: the Colab download path for the forecast CSV, which needs this branch merged to `main` before the raw URL resolves.

## Follow-ups

Left for the notebook authors, not blocked by this PR:

- **Bayesian optimization**: refreshed results differ a lot from the saved ones (TabPFN final gap mean 611 -> 2271, random search 6795 -> 4603, one repeat now favors random search). Separate PR for Klemens, together with removing the tqdm bar and softening the takeaway.
- **Relational predictions**: prints a pandas Styler repr instead of a table. Predates this change.

### CODE REVIEW

File: markdowns/bayesian_optimization.mdx
Comment (TuanaCelik) [inline markdowns/bayesian_optimization.mdx:1]: Can we remove the changes in this file for now and add in a separate one for Klemens to review? The outputs are vastly different..

File: scripts/rerun_notebooks.py
Comment (TuanaCelik) [inline scripts/rerun_notebooks.py:1]: Lgtm overall, just to check that I also undesrstood the readme: the 'rerun notebooks' script updating the models are manually triggered when we want to correct? 

We can also discuss this next week but until then I'm happy with this if I understood it correctly

**Comment by eliott-kalfon** [comment]
Something I would add:

- Results will change as some models have different behaviours. As the author of the large majority of these cookbooks, updating all of them as we release will represent a workload that does not match business impact (in my opinion)
- Happy to have these scripts in automation, but I think we have to be ready to ignore some of these changes
- I have changed the thinking notebook, adding a bit more depth to it with the `group_col` which should make it more useful

File: scripts/rerun_notebooks.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+SCRIPTS_DIR = ROOT / "scripts"
+COLAB_SHIM_DIR = SCRIPTS_DIR / "colab_shim"
+EXECUTOR = SCRIPTS_DIR / "_execute_notebook.py"
+WORK_DIR = ROOT / ".rerun"
+
+# Notebooks whose saved outputs come from a TabPFN model (hosted API or the
+# local ``tabpfn`` package) and therefore change with every model release.
+# The remaining notebooks train their own model or use a separate stack.
+MODEL_DEPENDENT_SLUGS = (
+    "quickstart",
```
Comment (eliott-kalfon) [inline scripts/rerun_notebooks.py]: Are we making this a bit brittle if we hard code lists in there?

**Review by eliott-kalfon — DISMISSED** [review]
LGTM, left a comment

File: markdowns/bayesian_optimization.mdx
Comment (psinger-prior) [reply markdowns/bayesian_optimization.mdx:1]: Done, the Bayesian optimization notebook is back to what is on main. The refreshed outputs and the prose fixes will go in a separate PR for Klemens.

File: scripts/rerun_notebooks.py
Comment (psinger-prior) [reply scripts/rerun_notebooks.py:1]: Yes, purely manual: someone runs it after a release, reviews the diff and opens a PR. Nothing triggers it automatically. We also took the section out of the README; the script's docstring documents usage and we handle each release by hand.

File: scripts/rerun_notebooks.py (hunk 1 above)
Comment (psinger-prior) [reply scripts/rerun_notebooks.py]: Agreed. The list is gone: the runner now picks every notebook whose pip install line includes tabpfn, tabpfn-client or tabpfn-extensions, which selects the same 11 and skips nanoTabPFN, TabPFN-Rel and the time-series notebook automatically. The only remaining list is a two-entry exclusion for the comparison cookbooks (tabpfn_35_vs_baselines, zomato_model_comparison), whose outputs must stay as recorded.

**Comment by psinger-prior** [comment]
The runner makes the mechanical rerun cheap; the plan is to trigger it manually per release and only touch prose where a conclusion flips, accepting number drift elsewhere. Thanks for the Thinking Mode revision, that closes that follow-up.

I do not think the rerunning is lots of work though.

**Review by eliott-kalfon — APPROVED** [review]
